src/protocols/secure_channel/CASEServer.cpp

```diff
@@ -46,76 +46,51 @@ CHIP_ERROR CASEServer::ListenForSessionEstablishment(Messaging::ExchangeManager
+CHIP_ERROR CASEServer::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, Messaging::ExchangeDelegate *& newDelegate)
 {
-    ReturnErrorCodeIf(ec == nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    if (mPairingSession.HasValue())
+        return CHIP_ERROR_NO_MEMORY;
+
+    CASESession * session = &mPairingSession.Emplace();
 
+    ChipLogProgress(Inet, "CASE Server allocated pairing: %p", session);
     // Setup CASE state machine using the credentials for the current fabric.
-    GetSession().SetGroupDataProvider(mGroupDataProvider);
+    session->SetGroupDataProvider(mGroupDataProvider);
     ReturnErrorOnFailure(
-        GetSession().ListenForSessionEstablishment(*mSessionManager, mFabrics, mSessionResumptionStorage, this,
-                                                   Optional<ReliableMessageProtocolConfig>::Value(GetLocalMRPConfig())));
-
-    // Hand over the exchange context to the CASE session.
-    ec->SetDelegate(&GetSession());
+        session->ListenForSessionEstablishment(*mSessionManager, mFabrics, mSessionResumptionStorage, this,
+                                               Optional<ReliableMessageProtocolConfig>::Value(GetLocalMRPConfig())));
+    newDelegate = session;
 
     return CHIP_NO_ERROR;
 }
 
-CHIP_ERROR CASEServer::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate)
-{
-    // TODO: assign newDelegate to CASESession, let CASESession handle future messages.
-    newDelegate = this;
-    return CHIP_NO_ERROR;
-}
-
-CHIP_ERROR CASEServer::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
-                                         System::PacketBufferHandle && payload)
-{
-    ChipLogProgress(Inet, "CASE Server received Sigma1 message. Starting handshake. EC %p", ec);
-    CHIP_ERROR err = InitCASEHandshake(ec);
-    SuccessOrExit(err);
-
-    // TODO - Enable multiple concurrent CASE session establishment
-    // https://github.com/project-chip/connectedhomeip/issues/8342
-    ChipLogProgress(Inet, "CASE Server disabling CASE session setups");
-    mExchangeManager->UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::CASE_Sigma1);
-
-    err = GetSession().OnMessageReceived(ec, payloadHeader, std::move(payload));
-    SuccessOrExit(err);
-
-exit:
-    if (err != CHIP_NO_ERROR)
-    {
-        Cleanup();
-    }
-    return err;
-}
-
-void CASEServer::Cleanup()
+void CASEServer::OnExchangeCreationFailed(Messaging::ExchangeDelegate * delegate)
 {
-    // Let's re-register for CASE Sigma1 message, so that the next CASE session setup request can be processed.
-    // https://github.com/project-chip/connectedhomeip/issues/8342
-    ChipLogProgress(Inet, "CASE Server enabling CASE session setups");
-    mExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::CASE_Sigma1, this);
-
-    GetSession().Clear();
+    mPairingSession.ClearValue();
```

Comment:
Should we do it? What if we're in the middle of a CASE and another CASE is about to be started but fails due to lack of free exchanges. I guess it shouldn't close an existing CASE, right?

Reply:
But still, what's exactly a scenario where this is useful?